# 이전 DNN 코드

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import optimizers
from tensorflow.keras.layers import Dense, Flatten, Input

import time

In [ ]:
(raw_train_x, raw_train_y), (raw_test_x, raw_test_y) = tf.keras.datasets.mnist.load_data()

train_x = raw_train_x/255.
test_x = raw_test_x/255.

train_y = raw_train_y
test_y = raw_test_y

In [ ]:

model = keras.Sequential()
model.add(Input((28,28)))
model.add(Flatten())
model.add(Dense(10, activation='relu'))
model.add(Dense(10, activation='relu'))
model.add(Dense(10, activation='softmax'))

model.compile(optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"])
model.summary()


model.fit(train_x, train_y, epochs=5, verbose=1, batch_size=128)


loss, acc = model.evaluate(test_x, test_y)
print("loss=",loss)
print("acc=",acc)

y_ = model.predict(test_x)
predicted = np.argmax(y_, axis=1)

print(predicted)


W0821 00:48:13.984249 140281580390272 deprecation.py:506] From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/init_ops.py:1251: calling VarianceScaling.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor


Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten (Flatten)            (None, 784)               0         
_________________________________________________________________
dense (Dense)                (None, 10)                7850      
_________________________________________________________________
dense_1 (Dense)              (None, 10)                110       
_________________________________________________________________
dense_2 (Dense)              (None, 10)                110       
Total params: 8,070
Trainable params: 8,070
Non-trainable params: 0
_________________________________________________________________
Epoch 1/5
60000/60000 [==============================] - 1s 20us/sample - loss: 0.7731 - acc: 0.7749
Epoch 2/5
60000/60000 [==============================] - 1s 19us/sample - loss: 0.3412 - acc: 0.9015
Epoch 3/5
60000/60000 [===========================

# CNN

In [1]:
import numpy as np
import matplotlib.pyplot as plt

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import optimizers
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPooling2D, Input, Reshape

import time

In [2]:
(raw_train_x, raw_train_y), (raw_test_x, raw_test_y) = tf.keras.datasets.mnist.load_data()

train_x = raw_train_x/255.
test_x = raw_test_x/255.

train_y = raw_train_y
test_y = raw_test_y

11501568/11490434 [==============================] - 0s 0us/step


In [3]:
#cnn 구조
model = keras.Sequential()
model.add(Input((28,28))) #입력층
#DNN에서 추가된 코드. ----------------
#(이 부분이 없으면 DNN)
#conv, maxpolling있는것(사용)이 CNN
model.add(Reshape((28,28,1))) # ADDED

#(3,3) : filter의 크기
# 32 : filter의 개수 
# (3,3)인 convolution filter적용해서 모두 32개 뽑음. padding은 입력과 출력사이즈 동일하게 same으로 맞춘다.
model.add(Conv2D(32, (3, 3), padding='same')) # ADDED
#maxpolling 사이즈 2x2
model.add(MaxPooling2D((2, 2))) # ADDED
model.add(Conv2D(64, (3, 3), padding='same')) # ADDED
model.add(MaxPooling2D((2, 2))) # ADDED
###-----추가된 코드 -----------------
model.add(Flatten())
model.add(Dense(10, activation='relu')) #히든 #cf ) 여기서 히든레이어 2개인 이유는 그냥임. 교재 만들면서 무작위로 만든 것.
model.add(Dense(10, activation='relu')) #히든
model.add(Dense(10, activation='softmax')) #출력층

#컴파일
model.compile(optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"])
model.summary()

#학습
model.fit(train_x, train_y, epochs=5, verbose=1, batch_size=128)

#평가
loss, acc = model.evaluate(test_x, test_y)
print("loss=",loss)
print("acc=",acc)

#예측
y_ = model.predict(test_x)
predicted = np.argmax(y_, axis=1)

print(predicted)
#dnn으로 했을 때는 92퍼 정확도, cnn으로 했을 때는 97퍼 정확도를 보이고 있음.

#conv 개수,필터 사이즈 몇개, 히든레이어 몇개, 노드 몇개, 등등 => 하이퍼 파라미터라고 함.
# 하이퍼 파라미터를 잘 조절해야 성능이 좋아짐. -> 이 작업을 리서처가함. 이 과정을 파인튠닝이라고함. -> 실무할때는 거의 안함.
#그럼 성능은 어케함? 연구자가 연구해놓은 결과들을 가져다가 씀.
#imagenet competition으로 가서 우수한 성능, 모델들이 있음. ex) vgg16

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 reshape (Reshape)           (None, 28, 28, 1)         0         
                                                                 
 conv2d (Conv2D)             (None, 28, 28, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 14, 14, 32)       0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 14, 14, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 7, 7, 64)         0         
 2D)                                                             
                                                                 
 flatten (Flatten)           (None, 3136)              0

In [ ]:
#ex) vgg16
#fileter 사이즈, 수 어케 결정? 결정 안함. 그냥 논문에 써 있는 것 가져다가 씀. 성능이 좋은 걸로

model.add(Conv2D(64,(3,3),padding = "same"))
model.add(Conv2D(64,(3,3),padding = "same"))
model.add(MaxPooling2D(2,2))
model.add(Conv2D(128,(3,3),padding = "same"))
model.add(Conv2D(128,(3,3),padding = "same"))
model.add(MaxPooling2D(2,2))
model.add(Conv2D(256,(3,3),padding = "same"))
model.add(Conv2D(256,(3,3),padding = "same"))
model.add(Conv2D(256,(3,3),padding = "same"))
model.add(Conv2D(256,(3,3),padding = "same"))
model.add(MaxPooling2D(2,2))
model.add(Conv2D(512,(3,3),padding = "same"))
model.add(Conv2D(512,(3,3),padding = "same"))
model.add(Conv2D(512,(3,3),padding = "same"))
model.add(MaxPooling2D(2,2))
model.add(Conv2D(512,(3,3),padding = "same"))
model.add(Conv2D(512,(3,3),padding = "same"))
model.add(Conv2D(512,(3,3),padding = "same"))
model.add(MaxPooling2D(2,2))
model.add(Flatten())
model.add(Dense(...))
model.add(Dense(...))
model.add(Dense(1000,activation = "softmax"))






필터의 값 같을 때 최대값을 가질 수 있음.

즉, 원본 이미지에 필토 모양 성분이 얼만큼 있는지를 출력하는 것이 convolution filter의 미가됨.

+) 필토 모야으이 자극이 있으면 그 결과가 최대가 된다, 필터 모양이 있는지 찾아낸다.

- MaxPoolin : 최대인 값을 뽑는 것.

conv2d 레이어의 파라매터 수는 다음에 의해서 결정된다.
```
파라매터 수 = (conv필터 가로 * conv 필터 세로 * 데이터 깊이 + bias 1개) * 필터 수
```
위 모델의 경우 파라매터 수 320, 18496은 다음과 같이 결정된다.

```
320 = (3*3*1+1) * 32

18496 = (3*3*32+1)*64
```

```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
conv2d (Conv2D)              (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 11, 11, 64)        18496     
```

# 영상 데이터 분류 CNN Template

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import optimizers
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPooling2D, Input, Reshape

import time

(raw_train_x, raw_train_y), (raw_test_x, raw_test_y) = tf.keras.datasets.mnist.load_data()

train_x = raw_train_x/255.
test_x = raw_test_x/255.

train_y = raw_train_y
test_y = raw_test_y

model = keras.Sequential()
model.add(Input((28,28)))
model.add(Reshape((28,28,1)))
model.add(Conv2D(32, (3, 3)))
model.add(MaxPooling2D((2, 2)))
model.add(Conv2D(64, (3, 3)))
model.add(MaxPooling2D((2, 2)))
model.add(Flatten())
model.add(Dense(10, activation='relu'))
model.add(Dense(10, activation='relu'))
model.add(Dense(10, activation='softmax'))


model.compile(optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"])
model.summary()


model.fit(train_x, train_y, epochs=5, verbose=1, batch_size=128)


loss, acc = model.evaluate(test_x, test_y)
print("loss=",loss)
print("acc=",acc)

y_ = model.predict(test_x)
predicted = np.argmax(y_, axis=1)

print(predicted)


11493376/11490434 [==============================] - 0s 0us/step
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
reshape (Reshape)            (None, 28, 28, 1)         0         
_________________________________________________________________
conv2d (Conv2D)              (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
flatten (Flatten)            (None, 1600)              0         
_________________________________________________________